shell/platform/linux/fl_renderer.cc

```diff
@@ -27,12 +48,46 @@ typedef struct {
   // was rendered
   bool had_first_frame;
 
-  GdkGLContext* main_context;
-  GdkGLContext* resource_context;
+  // Shader program.
+  GLuint program;
+
+  // Textures to render.
+  GPtrArray* textures;
 } FlRendererPrivate;
 
 G_DEFINE_TYPE_WITH_PRIVATE(FlRenderer, fl_renderer, G_TYPE_OBJECT)
 
+// Returns the log for the given OpenGL shader.
+static gchar* get_shader_log(GLuint shader) {
+  int log_length;
+  gchar* log;
+
+  glGetShaderiv(shader, GL_INFO_LOG_LENGTH, &log_length);
+
+  log = static_cast<gchar*>(g_malloc(log_length + 1));
```

Comment:
Here and below: consider documenting that the caller is responsible for freeing the returned string. Not super urgent since it's restricted to this translation unit, and all uses of these are using autofree, but kind to future readers :)

Reply:
In GLib this is somewhat implied - `gchar*` is allocated, and `const gchar*` is not. Wouldn't hurt to add to the doc though.